metrics/src/main/java/mobile/aerogear/org/metrics/MetricsService.java

```diff
@@ -0,0 +1,112 @@
+    private MetricsConfig config;
+
+    /**
+     * Get or create the client ID that identifies a device as long as the user doesn't
+     * reinstall the app or delete the app storage. A random UUID is created and stored in the
+     * application shared preferences.
+     *
+     * @param context Android app context
+     * @return String Client ID
+     */
+    private String getOrCreateClientId(Context context) {
+        SharedPreferences preferences = context
+            .getSharedPreferences(STORAGE_NAME, Context.MODE_PRIVATE);
+
+        String clientId = preferences.getString(STORAGE_KEY, null);
+        if (clientId == null) {
+            clientId = UUID.randomUUID().toString();
+
+            Log.d(LOG_TAG, "Generated a new client ID: " + clientId);
+
+            SharedPreferences.Editor editor = preferences.edit();
+            editor.putString(STORAGE_KEY, clientId);
+            editor.commit();
+        }
+
+        return clientId;
+    }
+
+    /**
+     * Get the version of the app itself
+     *
+     * @param context Android application context
+     * @return String version name
+     */
+    private String getAppVersion(Context context) {
+        try {
+            return context
+                .getPackageManager()
+                .getPackageInfo(context.getPackageName(), 0).versionName;
+        } catch (PackageManager.NameNotFoundException e) {
+            Log.e(LOG_TAG, e.getMessage());
+            return null;
+        }
+    }
+
+    /**
+     * Get the version of the SDK. This relies on gradle passing the version string
+     * to the metrics subproject
+     *
+     * @return String version name
+     */
+    private String getSdkVersion() {
+        return BuildConfig.VERSION_NAME;
+    }
+
+    private byte[] getMetricsData(Context context) throws JSONException {
+        JSONObject result = new JSONObject();
+        result.put("clientId", getOrCreateClientId(context));
+        result.put("appId", context.getPackageName());
+        result.put("appVersion", getAppVersion(context));
+        result.put("sdkVersion", getSdkVersion());
+
+        return result.toString().getBytes();
+    }
+
+    @Override
+    public void bootstrap(MobileCore core, ServiceConfiguration configuration) {
+        httpService = (HttpServiceModule) core.getService("http");
+        config = new MetricsConfig(configuration);
+    }
+
+    /**
+     * Send the metrics data to the server. The data is contained in a JSON object with the
+     * following properties: clientId, appId, sdkVersion and appVersion
+     *
+     * @param context Android application context
+     */
+    public void init(Context context) {
+        try {
+            // Send request to backend
+            HttpRequest request = httpService.newRequest();
+            request.post(config.getUri(), getMetricsData(context));
+            request.execute();
+            // TODO: response callback and response status logging
```

Comment:
@secondsun I tried
```
            final HttpResponse response = request.execute();
            response.onComplete(new Runnable() {
                @Override
                public void run() {
                    Log.d(LOG_TAG, "status: " + response.getStatus());
                }
            });
```
but that results in a null pointer exception when it's trying to get the status code from the response. What's the recommended way to use the http module in such a case? It looks like it's already executing the http requests on the network thread, so no need to use AsyncTask?